feat: add standalone admin api (#12179)
apache/apisix: docs/en/latest/deployment-modes.md

1. Kubernetes(k8s)：Declarative API that dynamically updates the routing rules with a full yaml configuration.
2. Different configuration centers: There are many implementations of the configuration center, such as Consul, etc., using the full yaml file for intermediate conversion.

### Modes

Now, we have two standalone running modes, file-driven and API-driven.

1. The file-driven mode is the kind APISIX has always supported.

    The routing rules in the `conf/apisix.yaml` file are loaded into memory immediately after the APISIX node service starts. At each interval (default: 1 second), APISIX checks for updates to the file. If changes are detected, it reloads the rules.

    *Note*: Reloading and updating routing rules are all hot memory updates. There is no replacement of working processes, since it's a hot update.

>     This requires us to set the APISIX role to data plane. That is, set `deployment.role` to `data_plane` and `deployment.role_data_plane.config_provider` to `yaml`.

Reviewer: Why is indentation added in front of these paragraphs? I don't think that's necessary.

Author: These text blocks must be in an ordered list, not aligned from the beginning of the line. The use of indentation accomplishes exactly that.

Author: ![image](https://github.com/user-attachments/assets/eed96ad7-50c7-43b9-9eff-44734a9663b5)

>     This requires we set the APISIX role to data plane. That is, set `deployment.role` to `data_plane` and `deployment.role_data_plane.config_provider` to `yaml`.

Reviewer 2: ```suggestion
    This requires us to set the APISIX role to data plane. That is, set `deployment.role` to `data_plane` and `deployment.role_data_plane.config_provider` to `yaml`.
```

    Refer to the example below:

    ```yaml
    deployment:
      role: data_plane
      role_data_plane:
        config_provider: yaml
    #END
    ```

    This makes it possible to disable the Admin API and discover configuration changes and reloads based on the local file system.

2. The API-driven is an emerging paradigm for standalone.

    The routing rules will be entirely in memory and not in a file, and it will need to be updated using the dedicated Standalone Admin API.

    I.e. we need to send an HTTP PUT request to this API containing the configuration in JSON or YAML format, which will flush the configuration used by each worker in the current APISIX instance.

>     Changes will overwrite the entire configuration and take effect immediately, no reboot is required as it is hot updated.

Reviewer 2: ```suggestion
    Changes will overwrite the entire configuration and take effect immediately without requiring a reboot, as it is hot updated.
```

>     This requires we set the APISIX role to traditional (since we need to start both the API gateway and the Admin API endpoint) and use the yaml config provider. That is, set `deployment.role` to `traditional`, and set `deployment.role_traditional.config_provider` to `yaml`

Reviewer 2: ```suggestion
    This requires us to set the APISIX role to traditional (since we need to start both the API gateway and the Admin API endpoint) and use the yaml config provider. That is, set `deployment.role` to `traditional` and `deployment.role_traditional.config_provider` to `yaml`.
```

    Refer to the example below:

    ```yaml
    deployment:
      role: traditional
      role_traditional:
        config_provider: yaml
    #END
    ```

>     This disables the local file source of configuration in favor of the API.When APISIX starts, it uses the empty configuration until you update it via the API.

Reviewer 2: ```suggestion
    This disables the local file source of configuration in favor of the API. When APISIX starts, it uses the empty configuration until you update it via the API.
```

    The following are API endpoints:

>     ```text

Reviewer 2: ```suggestion
    ```shell
```

    ## Update configuration
    ## The conf_version is not required, if it is not entered by the client, the current 10-digit epoch time is used by default.
    curl -X PUT http://127.0.0.1:9180/apisix/admin/configs?conf_version=1234 \
        -H "X-API-KEY: <apikey>"
        -H "Content-Type: application/json" ## or application/yaml
        --data-binary @config.json

    ## Get latest configuration
    curl -X GET http://127.0.0.1:9180/apisix/admin/configs
        -H "X-API-KEY: <apikey>"
        -H "Accept: application/json" ## or application/yaml
    ```

>     The update API checks our inputs and throws an error if they are incorrect. Returns `202 Accpeted` if the configuration was accepted, and returns the latest configuration version via the `X-APISIX-Conf-Version` header.

Reviewer 2: ```suggestion
    The update API validates the input and returns an error if it is invalid. If the configuration is accepted, it responds with a `202 Accepted` status and includes the latest configuration version in the` X-APISIX-Conf-Version` header.
```

    The get API also returns the version number via the `X-APISIX-Conf-Version` header, and returns a response body containing the configuration in a specific format as requested by the client `Accept` header.

>     These APIs follow the security requirements of the Admin API exactly, i.e. APIKey, TLS/mTLS, CORS, IP allowlist, etc., nothing special.

Reviewer 2: ```suggestion
    These APIs apply the same security requirements as the Admin API  — such as API key, TLS/mTLS, CORS, and IP allowlist — no changes or additions.
```

    The API accepts input in the same format as the file-based mode described above, although it also allows the user to input JSON instead of just YAML. The following example still applies. However, the API does not rely on the `#END` suffix because HTTP will guarantee input integrity.

>     *Note*: In this case, the Admin API based on etcd is not available, the configuration can only be flushed as a whole, rather than modifying a part of it, and the client needs to build request containing the complete new configuration to be sent to the API.

Reviewer 2: ```suggestion
    *Note*: In this case, the Admin API based on etcd is not available. The configuration can only be flushed as a whole, rather than modified partially, and the client must send a request containing the complete new configuration to the API.
```

>     *Note*: In this case, the Admin API based on etcd is not available, the configuration can only be flushed as a whole, rather than modifying a part of it, and the client needs to build request containing the complete new configuration to be sent to the API.

Reviewer 2: ```suggestion
    *Note*: In this case, the Admin API based on etcd is not available. The configuration can only be flushed as a whole, rather than modified partially, and the client must send a request containing the complete new configuration to the API.
```
